### PR TITLE
Us28

### PR DESCRIPTION
**Pull request by LelandCurtis**
- refactored all potential_revenue methods
-     potential revenue uses invoice_items quantity*unit_price, but only if they are associated with a successful transaction

- NOTE: if you want to calculate the revenue of an item you can't find all of the invoices connected to an item and sum the revenue of these invoices. They may have invoice_items that are not associated with that item. Same goes the other way, when finding revenue of an invoice, you can't sum all the revenue of the associated items. We need to focus on which invoice_items are are associated with our query. 
- You'll notice I wrote a potential_revenue method in invoice_items that I call from both items and invoices for this reason.

there are now potential revenue methods for:
- invoice_items
- items
- invoices
- invoice(potential_revenue_by_merchant)

I couldn't reuse the methods for the merchant.popular_items method. This needs to order by potential_revenue and I couldn't figure that out so I had to write a whole new query.

I also refactored the areas Eric was working on to work with these new methods. A lot of it was creating new data that fully tested it.

### CODE REVIEW

**Review by pleonar1 — APPROVED** [review]
looks good man

File: app/models/invoice.rb [hunk 1]
```diff
@@ -15,10 +15,14 @@ def merchant_items(merchant)
   end
 
   def merchant_invoice_items(merchant)
-    InvoiceItem.joins(:item).where( items: {merchant_id: merchant.id}).order("items.name asc")
+    invoice_items.joins(:item).where( items: {merchant_id: merchant.id}).order("items.name asc")
   end
```
Comment (EMielke76) [inline app/models/invoice.rb:19]: use the freebies. Love it.

File: app/models/invoice_item.rb [hunk 2]
```diff
@@ -7,4 +7,9 @@ class InvoiceItem < ApplicationRecord
 
 
   enum status: { "pending" => 0, :packaged => 1, "shipped" =>2 }
+
+  def self.potential_revenue
+    InvoiceItem.joins(invoice: :transactions).where(transactions: {result: 0})
+    .sum("invoice_items.quantity * invoice_items.unit_price")
```
Comment (EMielke76) [inline app/models/invoice_item.rb:13]: I was wondering there what `potential revenue` that kept cropping up was coming from. That's exactly what I was thinking we would need to do during the lesson w/Jamison this morning!

File: app/views/merchant_invoices/show.html.erb [hunk 3]
```diff
@@ -19,4 +19,4 @@
 <%end%>
 
 <h2>Total Potential Revenue</h2>
-<h4><%= money_honey(@invoice.potential_revenue(@merchant)) %></h4>
+<h4><%= money_honey(@invoice.potential_revenue_by_merchant(@merchant)) %></h4>
```
Comment (EMielke76) [inline app/views/merchant_invoices/show.html.erb:22]: That's slick as shit. Well done dude!

File: spec/models/invoice_item_spec.rb [hunk 4]
```diff
@@ -17,4 +17,37 @@
...
+        expect(invoice_items.potential_revenue).to eq(0)
+
+        # test for successful transactions
+        transaction_2 = create(:transaction, result: 0, invoice: invoice_1)
+        expect(invoice_items.potential_revenue).to eq(8000)
+
+        # test for multiple invoices with successful transactions
+        transaction_3 = create(:transaction, result: 0, invoice: invoice_2)
+        expect(invoice_items.potential_revenue).to eq(9000)
+      end
```
Comment (EMielke76) [inline spec/models/invoice_item_spec.rb:50]: This is about as robust a test I think I've ever seen. Great. Work.

File: spec/models/invoice_spec.rb [hunk 5]
```diff
@@ -51,13 +51,36 @@
...
+        item2 = create(:item_with_invoices, name: 'Car', merchant: merchant_2, invoices: [invoice1], invoice_item_quantity: 5, invoice_item_unit_price: 20000)
+        transaction_2 = create(:transaction, invoice: invoice1, result: 0)
+
+        # revenue associated with this invoice should not be included in potential revenue calcs.
+        invoice2 = create(:invoice)
+        item3 = create(:item_with_invoices, name: 'Plane', merchant: merchant_1, invoices: [invoice2], invoice_item_quantity: 3, invoice_item_unit_price: 33000)
+        item4 = create(:item_with_invoices, name: 'Yoyo', merchant: merchant_2, invoices: [invoice2], invoice_item_quantity: 5, invoice_item_unit_price: 77000)
+        transaction_3 = create(:transaction, invoice: invoice2, result: 0)
+
+        expect(invoice1.potential_revenue_by_merchant(merchant_1)).to eq(45000)
```
Comment (EMielke76) [inline spec/models/invoice_spec.rb:82]: Beauty. Really thorough testing.

File: spec/models/invoice_spec.rb [hunk 6]
```diff
@@ -51,13 +51,36 @@
...
+        expect(invoice1.potential_revenue_by_merchant(merchant_2)).to eq(100000)
```
Comment (EMielke76) [inline spec/models/invoice_spec.rb:83]: Beauty. Really thorough testing.

**Review by EMielke76 — APPROVED** [review]

**Pull request merged**